python/lsst/meas/base/_id_generator.py

```diff
@@ -0,0 +1,548 @@
+
+        class SomeTaskConfig(Config):
+            id_generator = VisitIdGeneratorConfig.make_field()
+
+        class SomeTask(Task):
+
+            ConfigClass = SomeTaskConfig
+
+            ...
+
+            def run(self, ..., data_id: DataCoordinate):
+                id_generator = self.config.apply(data_id)
+                catalog = id_generator.make_source_catalog(self.schema) ...
+
+    Code that wishes to instead unpack these record IDs to obtain the release
+    ID, data ID and counter value should use the same config (often loaded from
+    the ``Butler``) and pass a fully-expanded data ID identifying only a
+    particular ``skymap`` or ``instrument`` to `unpacker_from_config`::
+
+        config = butler.get("some_task_config")
+        catalog = butler.get("some_output_catalog", given_data_id)
+        unpacker = IdGenerator.unpacker_from_config(
+            config.id_generator, butler.registry.expandDataId(skymap="HSC"),
+        )
+        release_id, embedded_data_id, counter = unpacker(catalog[0]["id"])
+        assert embedded_data_id == given_data_id
+
+    This example is a bit contrived, as the ability to reconstruct the data ID
+    is really only useful when you don't have it already, such as when the
+    record ID is obtained from some further-processed version of the original
+    table (such as a SQL database), and in that context the right config to
+    load will not be obvious unless it has been carefully documented.
+
+    Simple instances of the base class that do not include a data ID may also
+    be constructed by calling the constructor directly::
+
+        id_generator = IdGenerator()
+
+    These IDs may not be unpacked, but they also don't need to be, because
+    they're just the per-catalog "counter" integer already.
+    """
+
+    # No deprecation decorator here because the type this method accepts is
+    # itself deprecated, so it's only going to be called be code paths that
+    # will go away when the deprecation turns into a removal, and which already
+    # warn.
+    @staticmethod
+    def _from_exposure_id_info(exposure_id_info: ExposureIdInfo) -> IdGenerator:
+        """Construct a new ID generator from the object this class supersedes.
+
+        This method is deprecated along with the type it accepts; it's provided
+        only as a temporary helper to aid in the transition from
+        `lsst.obs.base.ExposureIdInfo` to `IdGenerator`.
+        """
+        return _ExposureIdInfoIdGenerator(exposure_id_info)
+
+    @property
+    def catalog_id(self) -> int:
+        """The integer identifier for the full catalog with this data ID, not
+        just one of its rows (`int`).
+
+        This combines the packed data ID and release ID, but not the
+        counter.
+        """
+        return 0
+
+    def __str__(self) -> str:
+        """Return a human-readable representation of the data ID (or a note
+        about its absence) for use in log and error messages.
+        """
```

Comment:
Is the dataId the only thing we want to include in these `str`, or should it say something about the number of bits, too?

Reply:
I was envisioning `str` here as being really for logs and exceptions about the catalog, not logs and exceptions about the IdGenerator itself.  There are a lot of tasks whose `run` methods take an `idGenerator` and that effectively abstracts over whether there is a data ID at all.

Reply:
We discussed this point on one of the other PRs; using the id generator to access a stringified dataId seems like a bit of an abuse of the system. Maybe just note in the `IdGenerator` docs that we use it as such?